### ISSUE TITLE
Forward AWS credentials

### ISSUE THREAD

**Issue by mrocklin** [coiled/feedback#6]
I'm thinking a bit about how to handle AWS credentials.  There are two approaches here:

1.  We maintain an IAM role with all of the policies necessary as part of the cluster type definition in our database.  Users configure this either through some web UI or the Python API and we ship it out to workers as part of the task definition.

2.  We collect credentials from the user's machine, and distribute those to the workers

From a developer and usability standpoint I prefer the second approach.  The burden of maintaining credentials is passed to the user, so we don't have to worry about it and so it's familiar to them.  Also the user doesn't have to interact with whatever system we make to create IAM roles, which is both a reinvention of AWS technology and also likely to be a speedbump to non-expert users.

However, I'm curious about how people feel about transmitting their AWS credentials to other machines.  This would all be done over secure connections of course, but it may be that people want compute machinery to be less privileged than personal machines.

If this was the case, they could craft a role of their own, and pass those credentials when making the cluster.  A sample API looks like this.

```python
cluster = CoiledCluster(..., credentials=True)  # gather local credentials, pass to workers.  The default

cluster = CoiledCluster(..., credentials={"access_key": ..., "secret_key": ...)  # pass in credentials for pre-defined role
```

@joeschmid , as someone who cares both about usability and security I'd be curious about your perspective here if you have time.

**Comment by mrocklin** [comment]
Perhaps we would use something like `credentials={"session_token": ...}` when we pass around local credentials that we've gathered automatically.

**Comment by joeschmid** [comment]
@mrocklin I liked the way you laid out those options and thought about the tradeoffs. My two cents goes like this:

- In terms of AWS and infosec best practices, # 1 is the preferred approach. Obviously IAM roles avoid the need to pass around credentials and the risks that come with that. (The [AWS best practices doc](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html) says directly: "To provide credentials to the application in a secure way, use IAM roles.")
- However, many users will not have significant experience with IAM roles, but will have credentials.

My sense is that most veteran AWS users will have already adopted approach # 1 and prefer that, but for new users approach # 2 will be more straightforward, require less configuration/AWS knowledge, and will get them up and running faster.

TL;DR: both, please!

**Comment by mrocklin** [comment]
So for option 1 you'd be comfortable with a Python API like the following?

```python
cluster = CoiledCloud(n_workers=100, role="some-iam-role-arn")
```

And, as a corollary, you would be comfortable giving our service (running inside your VPN) the authority to create tasks with these roles?

**Comment by ssanderson** [comment]
> `CoiledCluster(..., credentials=...)`

One implication of this API is that it means credentials are scoped to a cluster, rather than a user. Is it possible (or might it be possible in the future) for multiple users to connect to the same coiled cluster? Or is the expectation that each user creates and works with their own clusters? I could imagine there being efficiencies to be had from sharing resources in some cases.

If it's possible for multiple users to connect to the same cluster, than it seems like you wouldn't want the cluster's permissions to be tied to the credentials of any particular user, so you'd either have to make an IAM role, or you'd have to create credentials tied to the capabilities of the cluster and remember to use those (which seems more cumbersome than using a role).

**Comment by ssanderson** [comment]
Also, :+1: for thinking about this problem. I sat down today to try working on a problem where I wanted to load some data from S3, and the first question I ran into was "how do I manage access to that?".

**Comment by mrocklin** [comment]
> One implication of this API is that it means credentials are scoped to a cluster, rather than a user. Is it possible (or might it be possible in the future) for multiple users to connect to the same coiled cluster? Or is the expectation that each user creates and works with their own clusters? I could imagine there being efficiencies to be had from sharing resources in some cases.

Agreed.  I think that the common case deployment is one-cluster-per-user.  (this is in no way enforced today by the way).  In this common case, I think that the most desired policy will be to mimic that user everywhere on the cluster.  Do you agree?

Once we start sharing clusters then we would need to stop this policy and switch to something else, like providing IAM roles, or passing up a different set of credentials.

> Also, +1 for thinking about this problem. I sat down today to try working on a problem where I wanted to load some data from S3, and the first question I ran into was "how do I manage access to that?".

How do you want to manage access to that?

**Comment by ssanderson** [comment]
> I think that the common case deployment is one-cluster-per-user.

I think this depends on how easy it is to share results between users in different clusters. If people are only using their clusters as ephemeral compute resources and storing anything persistent/shared in S3 (or similar), then I think the one-cluster-per-user model makes sense. If people are persisting results to disk in the cluster itself, then I could imagine there being some pull toward wanting to share clusters. Similarly, if people are doing analyses with large shared resources that are hard to access in a cloud-native way (e.g., big HDF5 files on disk), I could imagine wanting to co-locate them to reduce the number of copies that need to be made of those files. This is a place where you probably have better intuitions than I do w/r/t how dask distributed users organize their workflows.

>  In this common case, I think that the most desired policy will be to mimic that user everywhere on the cluster. Do you agree?

I definitely agree that in the one-cluster-per-user case, the most common desire will be to have the cluster's permissions match those of the user. In more security-sensitive contexts, I'd expect another common desire to be for the cluster to only have a subset of the user's permissions. Thinking about my credentials at Quantopian, for example, I had access to a variety of AWS-managed resources, some of which were considered more sensitive than others, and some of which were more likely to be useful in a dask-based analysis than others. For something like that, I could imagine wanting to limit the permissions of my cluster to a subset of the resources my user had access to. My guess is that anyone with that level of complexity in their desires would want to be using IAM roles anyway though.

> How do you want to manage access to that?

For my current use-case (poking around at hobby projects to learn more about dask), the most **convenient** thing would be for dask to pick up credentials from my local ~/.aws/credentials and have the permissions associated with those credentials be "added" to the cluster. I'd be a little bit nervous about having my credentials stored persistently in the cluster though. One way to mitigate that would be to generate [temporary credentials](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp.html) from my access keys, and use those, though it's unclear what the re-auth flow would look like for that. Another option (I **think**, I'm not quite totally following the docs for this) might be for coiled to act as an [identity provider](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) for my AWS account.

**Comment by shughes-uk** [comment]
Done 😉
